compiler/executable/match_/planner/plan.rs

```diff
@@ -693,87 +696,32 @@ impl<'a> ConjunctionPlanBuilder<'a> {
                     plan.cumulative_cost.chain(plan.ongoing_step_cost),
                     plan.heuristic
                 );
-
+                debug_assert!(extension_heap.is_empty());
+                // Add best k extensions from this plan to new_plan_heap (k = extension_width)
                 for extension in plan.extensions_iter(&self.graph)? {
                     if extension.is_trivial(&self.graph) {
-                        event!(
-                            Level::TRACE,
-                            "{INDENT:12}Stash {:?} = {} <-- cost: {:?} heuristic: {:?}",
-                            extension.pattern_id,
-                            self.graph.elements[&VertexId::Pattern(extension.pattern_id)],
-                            extension.step_cost.cost,
-                            extension.heuristic
-                        );
-                        let mut plan = plan.clone();
-                        plan.add_to_stash(extension.pattern_id, &self.graph);
-                        if new_plans_heap.len() < beam_width {
-                            new_plans_heap.push(plan);
-                        } else if let Some(top) = new_plans_heap.peek() {
-                            if plan < *top {
-                                new_plans_heap.pop();
-                                new_plans_heap.push(plan);
-                            }
-                        }
                         extension_heap.clear();
+                        extension_heap.push(Reverse(extension));
                         break;
-                    }
-
-                    if extension_heap.len() < extension_width {
-                        extension_heap.push(extension);
-                    } else if let Some(top) = extension_heap.peek() {
-                        if extension < *top {
-                            extension_heap.pop();
-                            extension_heap.push(extension);
-                        }
```

Comment:
I could keep this way of doing things, but figured using two unbounded size min-heaps is less confusing than one unbounded-min and one bounded-max.

Reply:
The fact that the extension heap grows without bounds makes me a little uneasy. The number of extensions _per prefix_ is ~ N, and the beam width is tied to N (although capped), so we're looking at an O(N^2) space extension heap here.

Reply:
Valid. I can revert this if space is a concern. I don't think we have much of a time-cost because we generate the extensions regardless.

Reply:
@dmitrii-ubskii To clarify: We drain the extension heap after each plan. So it's still limited to `O(N)` space.
The new plans heap will have a size of `O(beam_width * extension_width)`  overhead. This is indeed `O(N^2)`. 

I don't want to use the old approach of capping this to `beam_width` at all times because 
1. We can't check the hash before inserting to `new_plans_heap` as they're not added in order of cost.
2. Since we check the hash when draining the `new_plans_heap`, we may rule out candidates and be left with less than `beam_width` candidates in the next step.